### PR TITLE
Add scopes support to OpenGroup factory

### PR DESCRIPTION
**Pull request by seanh**
Usage:

By default `factories.OpenGroup()` will return a group with a random
number of randomly generated scopes:

    >>> group = factories.OpenGroup()
    >>> group.scopes
    [<GroupScope http://martinez-shepherd.com/>,
     <GroupScope https://prince-ochoa.com/>]

You can specify how many randomly generated scopes you want a group to
have:

    >>> group = factories.OpenGroup(scopes=5)
    >>> len(group.scopes)
    5

You can create an open group with no scopes:

    >>> group = factories.OpenGroup(scopes=0)
    >>> group.scopes
    []

(passing `scopes=[]` will also work).

If you have some `GroupScope` objects you can create an open group with
a particular list of scopes or a list of one scope:

    group = factories.OpenGroup(scopes=[my_scope])

Private groups created with `factories.Group()` will have no scopes.

Incidentally this change also means that when you create a scope using
`factories.GroupScope()` and an open group is randomly generated for
that scope to belong to, a random number of _other scopes_ will also be
generated and added to the group's scopes:

    >>> scope = factories.GroupScope()
    <GroupScope https://www.jacobs-frazier.com/>

    >>> scope.group.scopes
    [<GroupScope https://www.payne-owens.info/>,
     <GroupScope https://www.jacobs-frazier.com/>]

### CODE REVIEW

File: tests/common/factories/group_scope.py [hunk 1]
```diff
@@ -7,12 +7,11 @@
...
-from .group import OpenGroup
 
 
 class GroupScope(ModelFactory):
     class Meta:
         model = models.GroupScope
 
     origin = factory.Faker('url')
-    group = factory.SubFactory(OpenGroup)
+    group = factory.SubFactory('tests.common.factories.OpenGroup')
```
Comment (seanh) [inline tests/common/factories/group_scope.py]: This is to avoid a circular import

**Comment by lyzadanger** [comment]
Thanks for getting this in place so fast! This looks good and helpful and flexible. The one thing that gives me pause is the default-creation-of-a-random-number-of-scopes — I worry that this automatic feature may cause some confusion later on in testing, especially when we have unscoped open groups floating around frequently. That is, I wonder if 0 scopes may make more sense as a default?

I have a hunch that the most common testing use cases in the near future might be:

1. Create an open group with a particular scope or scopes. This is what we'll be testing in the near future as we start filtering on scope.
2. Create an open group for which you don't care about scope. Tests in this vein already exist (e.g. in `ListGroupsService` and pending `session` tests).

This pattern is the one I'll be using soonest, naturally:

```
group = factories.OpenGroup(scopes=[my_scope])
```

What's is `my_scope`, though, in this case? It is a factory-created `GroupScope`? As you mentioned, if you create a `GroupScope` using `factories`:

`>>> scope = factories.GroupScope()`

It will automatically create a group to hold it and own it. However, that's chicken-before-egg if I want to do:

```
group = factories.OpenGroup(scopes=[my_scope])
```

as (another) group for that `my_scope` will have already been created.

tl;dr : I feel like a default of either 0 or 1 scopes automatically created for an open group may be less "magical", and I'm not clear on how to create a scoped group in test without side-effect creating an additional group.

**Comment by seanh** [comment]
Hmm, yeah, I think defaulting the scopes to 0 is probably best, so create an unscoped group (which we will actually have in production one day) by default, because that'll resolve the chicken and egg situation.

Currently if your test wants to get hold of a `Group` and a `GroupScope` that are related your options are:

```python
scope = factories.GroupScope()
open_group = scope.group
```

Or vice-versa. However if you're calling `GroupScope()` for example you can't pass arguments through to the `OpenGroup()` subfactory that it will call if you want to control that. You can modify the object's properties after creation of course. I think there may be a way to implement this pass-through but not sure.

So if you do want to call both factories yourself to specify args to each then you can either do this:

```python
scope = factories.GroupScope(...)
group = factories.OpenGoup(scopes=[scope], ...)
```

That will end up _moving_ `scope` from whatever automatically created group it was added to in the first place onto `group` and leaving an unscoped open group in the db.

To avoid leaving an unscoped group in the db you would do:

```python
scope = factories.GroupScope(group=None, ...)
group = factories.OpenGoup(scopes=[scope], ...)
```

Or

```python
group = factories.OpenGoup(scopes=[], ...)
scope = factories.GroupScope(group=group, ...)
```

Both of which seem awkward. Or just create a scope directly without using the factory:

```python
scope = models.GroupScope(...)
group = factories.OpenGoup(scopes=[scope], ...)
```

But changing `factories.OpenGroup()` to 0 scopes by default would fix it to this:

```python
group = factories.OpenGoup(...)
scope = factories.GroupScope(group=group, ...)
```

If you call `factories.GroupScope()` that will still by default auto create a group for the scope to belong to unless you pass `group=something` (where `something` can also be `None`). _This_ probably makes sense though because an ungrouped scope isn't actually valid (will crash if you try to flush the db session).

**Comment by lyzadanger** [comment]
> But changing factories.OpenGroup() to 0 scopes by default would fix it to this:

```
group = factories.OpenGoup(...)
scope = factories.GroupScope(group=group, ...)
```

This looks fine to me as a flow and I feel better with the default-0-scopes approach. Make it so!

**Comment by seanh** [comment]
Even with the fix though it's a bit odd that if a particularly-scoped group is what you want, then you have to first create an unscoped group and then create a scope and pass the group to the scope constructor:

```python
group = factories.OpenGoup(...)
scope = factories.GroupScope(group=group, ...)
```

Seems backwards. You can't do it the other way round, this will create an extra group in the db:

```python
scope = factories.GroupScope(...)
group = factories.OpenGoup(scopes=[scope], ...)
```

**Comment by seanh** [comment]
Actually you _will_ be able to do this:

```python
scope = factories.GroupScope(group=None)
group = factories.OpenGoup(scopes=[scope], ...)
```

So that's fine I guess.

**Comment by lyzadanger** [comment]
> Actually you will be able to do this:

```
scope = factories.GroupScope(group=None)
group = factories.OpenGoup(scopes=[scope], ...)
```

Even better (more natural)!

**Comment by seanh** [comment]
> > Actually you will be able to do this:
> > 
> > ```
> > scope = factories.GroupScope(group=None)
> > group = factories.OpenGoup(scopes=[scope], ...)
> > ```
> 
> Even better (more natural)!

Actually that doesn't work unfortunately, it crashes for annoying reasons. But you _can_ do it with `build()` though:

```python
scope = factories.GroupScope.build(group=None)
group = factories.OpenGoup(scopes=[scope], ...)
```

@lyzadanger I've force-pushed the branch with the change to default to 0 and updated the commit message.
